rust/xaynet-mobile/src/ffi/participant.rs

```diff
@@ -345,3 +362,143 @@ pub unsafe extern "C" fn xaynet_ffi_participant_set_model(
         ERR_SETMODEL_MODEL
     }
 }
+
+/// Return the latest global model from the coordinator.
+///
+/// - `buffer` is the array in which the global model should be copied.
+/// - `data_type` specifies the type of the model weights (see [`DataType`]). The C header
+///   file generated by this crate provides an enum corresponding to the parameters: `DataType`.
+/// - `len` is the number of weights the model has
+///
+/// # Return Value
+///
+/// - [`OK`] if the model is set successfully
+/// - [`ERR_NULLPTR`] if `participant` or the `buffer` is NULL
+/// - [`GLOBAL_MODEL_NONE`] if no model exists
+/// - [`ERR_GLOBALMODEL_IO`] if the communication with the coordinator
+/// - [`ERR_GLOBALMODEL_DATATYPE`] if the datatype is invalid
+/// - [`ERR_GLOBALMODEL_LEN`] if the length of the buffer does not match the length of the model
+/// - [`ERR_GLOBALMODEL_CONVERT`] if the conversion of the model failed
+///
+/// # Note
+///
+///    It is **not** guaranteed, that the model configuration returned by
+///    [`xaynet_ffi_participant_model_config`] corresponds to the configuration of
+///    the global model. This means that the global model can have a different length / data type
+///    than it is defined in model configuration. That both model configurations are the same is
+///    only guaranteed if the model config **never** changes on the coordinator side.
+///
+/// # Safety
+///
+/// 1. When calling this method, you have to ensure that *either* the pointer is NULL
+///    *or* all of the following is true:
+///    - The pointer must be properly [aligned].
+///    - It must be "dereferencable" in the sense defined in the [`::std::ptr`] module
+///      documentation.
+/// 2. If `len` or `data_type` do not match the model in `buffer`, this method will
+///    result in a buffer over-read.
+#[no_mangle]
+pub unsafe extern "C" fn xaynet_ffi_participant_global_model(
+    participant: *mut Participant,
+    buffer: *mut c_void,
+    data_type: c_uchar,
+    len: c_uint,
+) -> c_int {
+    let participant = match unsafe { participant.as_mut() } {
+        Some(participant) => participant,
+        None => return ERR_NULLPTR,
+    };
+
+    if buffer.is_null() {
+        return ERR_NULLPTR;
+    }
+
+    let global_model = match participant.global_model() {
+        Ok(Some(model)) => model,
+        Ok(None) => return GLOBAL_MODEL_NONE,
+        Err(_) => return ERR_GLOBALMODEL_IO,
+    };
+
+    let data_type = match DataType::try_from(data_type) {
+        Ok(data_type) => data_type,
+        Err(_) => return ERR_GLOBALMODEL_DATATYPE,
+    };
+
+    let len = len as usize;
+    if len != global_model.len() {
+        return ERR_GLOBALMODEL_LEN;
+    }
+
+    match data_type {
+        DataType::F32 => {
```

Comment:
Not sure, bit maybe a macro would help making this a bit shorter.

Reply:
That is a good idea👍